File: src/main/java/info/opensigma/module/Module.java
package info.opensigma.module;

import info.opensigma.OpenSigma;
import info.opensigma.setting.owner.SettingOwner;
import info.opensigma.system.INameable;

public class Module implements INameable {

    protected final SettingOwner settings;

    public final String name, description;
    private boolean enabled;
    public int key;

    public Module(String name, String description, int key) {
        this.name = name;
        this.description = description;
        this.enabled = false;
        this.key = key;

        this.settings = new SettingOwner(this);

        OpenSigma.getInstance().bindManager.add(() -> key, this::toggle);
    }

    public Module(String name, String description) {
        this(name, description, 0);
    }

    public final void init() {
        this.settings.onMinecraftLoad();
    }

    public final void toggle() {
        setEnabled(!enabled);
    }

    public final void setEnabled(boolean enabled) {
        if (enabled)
            enable();
        else
            disable();
    }

    public final void enable() {
        if (this.enabled)
            return;

        onEnable();

        OpenSigma.getInstance().eventBus.subscribe(this);

        enabled = true;
    }

    public final void disable() {
        if (!this.enabled)
            return;

        enabled = false;

        OpenSigma.getInstance().eventBus.subscribe(false);

        onDisable();
    }

    protected void onEnable() { }

    protected void onDisable() { }

    @Override
    public String getName() {
        return name;
    }
}

File: src/main/kotlin/info/opensigma/module/Module.kt
package info.opensigma.module

import info.opensigma.setting.owner.SettingOwner
import info.opensigma.system.INameable
import info.opensigma.system.IMinecraft
import info.opensigma.OpenSigma
import info.opensigma.module.data.ModuleCategory

open class Module(
    override val name: String,
    val description: String,
    var key: Int = 0,
    val category: ModuleCategory = ModuleCategory.MISC
) : INameable, IMinecraft {

    protected val settings: SettingOwner<Any> = SettingOwner(this)
    private var enabled: Boolean = false

    init {
        OpenSigma.instance.bindManager.add({ key }, this::toggle)
    }

    fun toggle() {
        setEnabled(!enabled)
    }

    fun setEnabled(enabled: Boolean) {
        if (enabled) enable() else disable()
    }

    fun enable() {
        if (this.enabled) return
        enabled = true
        onEnable()
        OpenSigma.instance.eventBus.subscribe(this)
    }

    fun disable() {
        if (!this.enabled) return
        enabled = false
        OpenSigma.instance.eventBus.subscribe(false)
        onDisable()
    }

    @Suppress("EmptyMethod")
    protected open fun onEnable() {}

    @Suppress("EmptyMethod")
    protected open fun onDisable() {}
}